# Tensorflow

En esta práctica se realizará un ejercicio usando la librería de inteligencia artificial `tensorflow`, en un documento interactivo Jupyter Notebook usando Python 3.8.8.

A continuación se construirá una red neuronal artificial la cual será entrenada para identificar caracteres escritos a mano. Primero importaremos la librería:

In [1]:
import tensorflow as tf

Y cargamos el conjunto de datos MNIST, el cual contiene las imágenes de caracteres escritos a mano y será lo que usaremos tanto para entrenar como para probar la red neuronal:

In [3]:
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0

Con la API de Keras podemos usar todas las características de Tensorflow de manera sencilla, pues para construir la red neuronal basta con agregar capa por capa de manera secuencial, esto lo logramos con unas cuantas líneas de código:

In [12]:
model = tf.keras.models.Sequential([
    # capa de entrada
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    # capa de procesamiento
    tf.keras.layers.Dense(128, activation='relu'),
    # capa para prevenir sobreentrenamiento
    tf.keras.layers.Dropout(0.2),
    # capa de salida
    tf.keras.layers.Dense(10),
    # capa para calcular probabilidades
    tf.keras.layers.Softmax()
])

Ahora pasaremos a entrenar el modelo con una parte del conjunto de datos:

In [14]:
model.compile(
    optimizer='adam',
    # función de pérdida, calcula el logaritmo negativo de una predicción verdadera
    loss=tf.keras.losses.SparseCategoricalCrossentropy(),
    metrics=['accuracy']
)
# 10 iteraciones
model.fit(x_train, y_train, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 9s 3ms/step - loss: 0.4822 - accuracy: 0.8579
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1488 - accuracy: 0.9563
Epoch 3/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1071 - accuracy: 0.9678
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0840 - accuracy: 0.9727
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0743 - accuracy: 0.9765
Epoch 6/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0616 - accuracy: 0.9801
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0550 - accuracy: 0.9816
Epoch 8/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0483 - accuracy: 0.9846
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0429 - accuracy: 0.9853
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.041

Después de diez iteraciones, el modelo reportó una precisión de 98.7% sobre el conjunto de datos de entrenamiento. Ya que entrenamos a la red neuronal, pasaremos a utilizar un método para evaluar qué tanto podemos confiar en las predicciones de nuestro modelo:

In [15]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 1s - loss: 0.0742 - accuracy: 0.9793


[0.07423324882984161, 0.9793000221252441]

Esta evaluación nos indica que la precisión real, es decir, después de haber probado el modelo con nuevos datos que nunca había visto, es de 97.9%, lo cual es normal que sea menor que la precisión de entrenamiento pues queremos ver la eficiencia de la red neuronal al predecir datos totalmente nuevos.